Capabilities 2020 chapter (#1382)
HTTPArchive/almanac.httparchive.org: src/content/en/2020/capabilities.md

---
part_number: II
chapter_number: 13
title: Capabilities
description: Capabilities chapter of the 2020 Web Almanac covering brand-new, powerful web platform APIs.
authors: [christianliebel]
reviewers: [logicalphase,tomayac]
analysts: [tomayac]
translators: []

> discuss: 1981

Reviewer: We have the 2020 discussion ids now and [your own is 2049](https://discuss.httparchive.org/t/chapter-13-capabilities/2049):

```suggestion
discuss: 2049
```

> discuss: 2049

Reviewer 2: FYI here's the URL for this discussion thread: https://discuss.httparchive.org/t/chapter-13-capabilities/2049

I'd recommend "watching" it so you can be notified of comments by readers.

Thanks to the [StorageManager API](https://developer.mozilla.org/en-US/docs/Web/API/StorageManager) (part of the [WHATWG Storage Living Standard](https://storage.spec.whatwg.org/#storagemanager)), browsers no longer behave like a black box in that regard: This API allows developers to estimate the remaining space available, and to opt-in to [persistent storage](https://web.dev/persistent-storage/), meaning that the browser will not clear a website's data when disk space is low. Therefore, the API introduces a new `StorageManager` interface on the `navigator` object, currently available on Chrome, Edge, and Firefox.

### Estimate the available storage

Developers can estimate the available storage by calling `navigator.storage.estimate()`. This method returns a promise resolving to an object with two properties: `usage` shows the number of bytes used by the application and `quota` contains the maximum number of bytes available.

{{ figure_markup(
  image="storage_manager_api_estimate.png",

>   alt="Percent of top pages using the estimate method of the StorageManager API"

Reviewer: Missing a comma here and in a couple of other places like this, which is causing the test failures. Could you correct?

```suggestion
  alt="Percent of top pages using the estimate method of the StorageManager API",
```

Reviewer 2: @christianliebel I was testing locally and went ahead and committed these fixes

featured_quote: The state of web capabilities in 2020 is healthy, as new powerful APIs regularly ship with new releases of Chromium-based browsers. The first APIs have already made their way into other browsers as well.
featured_stat_1: 0.0006%
featured_stat_label_1: (Chrome) Page loads using Async Clipboard API
featured_stat_2: 0.49%
featured_stat_label_2: Sites using Storage Manager API
featured_stat_3: 363
featured_stat_label_3: Sites allow installing related apps
published: 2020-11-09T00:00:00.000Z
last_updated: 2020-11-09T00:00:00.000Z
unedited: true
---

## Introduction

[Progressive Web Apps](./pwa) (PWA) are a cross-platform application model based on web technology. With the help of Service Workers, these applications run even when the user is offline. The Web App Manifest allows users to add a PWA to their home screen or program list. When opened from there, a PWA appears as a native application. However, PWAs can only use the functions and capabilities that are exposed through web platform APIs. Arbitrary native interfaces cannot be called, leaving a gap between native applications and web apps.

> The [Capabilities Project](https://www.chromium.org/teams/web-capabilities-fugu), informally also known as Project Fugu, is a cross-company effort by Google, Microsoft, and Intel to bridge the gap between web and native. To do so, the Chromium contributors implement new APIs exposing capabilities of the operating system to the web, while maintaining user security, privacy, and trust. These capabilities include, but are not limited to:

Reviewer 2: Can you talk about why it's important to bridge the gap between web and native? I think the answer has profound implications on the future of the web. Maybe it's too big a question to answer here, but it's something I generally don't see discussed enough. This chapter and Mobile Web are probably the best places to have that discussion, even briefly.

Author: Added a brief sentence. Not sure if this suffices. 😇

The codename of the Capabilities project is named after a Japanese dish: Correctly prepared, the meat of the blowfish is a special taste experience. If prepared incorrectly, however, it can be fatal. The powerful APIs of Project Fugu are extremely exciting for developers. However, they can affect the security and privacy of the user. Therefore, the Fugu team pays special attention to these issues: For instance, new interfaces require the website to be sent over a secure connection (HTTPS). Some of them require a user gesture, such as a click or key press, to prevent fraud. Other capabilities require explicit permission by the user. Developers can use all APIs as a progressive enhancement: By feature detecting the APIs, applications won't break in browsers lacking support for those capabilities. In browsers that support them, users can get a better user experience. This way, web apps [progressively enhance](https://web.dev/progressively-enhance-your-pwa/) according to the user's particular browser.

> This chapter gives an overview of various modern web APIs, and the state of web capabilities in 2020 based on usage data by the HTTP Archive. Due to technical limitations, the HTTP Archive only has data available for APIs that require neither permission, nor a user gesture. Where no data is available, the percentage of page loads in Google Chrome according to [Chrome Platform Status](https://chromestatus.com/metrics/feature/timeline/popularity) will be shown instead. Since some interfaces are brand-new and their usage is still very low, the statistics are not necessarily meaningful. Unlike most chapters, stats will be presented as the absolute numbers of pages rather than relative percentages, where available. However, in many cases trends can still be read from the data. Unless otherwise noted, the APIs are only available in Chromium-based browsers, and their specifications are in the early stages of standardization.

Reviewer 2: One other point that could be good to make here is that measuring low adoption just means that we have a baseline from which to compare future growth. Future annual editions of this chapter can look back to these stats to see how much the APIs have matured and improved their adoption.

Thanks to the [StorageManager API](https://developer.mozilla.org/en-US/docs/Web/API/StorageManager) (part of the [WHATWG Storage Living Standard](https://storage.spec.whatwg.org/#storagemanager)), browsers no longer behave like a black box in that regard: This API allows developers to estimate the remaining space available, and to opt-in to [persistent storage](https://web.dev/persistent-storage/), meaning that the browser will not clear a website's data when disk space is low. Therefore, the API introduces a new `StorageManager` interface on the `navigator` object, currently available on Chrome, Edge, and Firefox.

### Estimate the available storage

Developers can estimate the available storage by calling `navigator.storage.estimate()`. This method returns a promise resolving to an object with two properties: `usage` shows the number of bytes used by the application and `quota` contains the maximum number of bytes available.

{{ figure_markup(
  image="storage_manager_api_estimate.png",
  alt="Number of top pages using the estimate method of the StorageManager API",
  caption="Number of top pages using the estimate method of the StorageManager API.",

>   description="Chart of the usage of StorageManager API's estimate method, based on the percent of pages monitored by HTTPArchive. It compares the usage on mobile and desktop devices. It shows a linear growth on the desktop, while it shows a hockey stick growth for mobile devices. In October, approximately 0.5 percent of both mobile and desktop pages make use of it.",

Reviewer: Need to change this now graph is using absolute numbers:

```suggestion
  description="Chart of the usage of StorageManager API's estimate method, based on the number of pages monitored by HTTPArchive. It compares the usage on mobile and desktop devices. It shows a linear growth on the desktop, while it shows a hockey stick growth for mobile devices. In October, approximately 34,000 mobile sites and 27,000 desktop sites make use of it.",
```

### Opt-in to persistent storage

There are two categories of web storage: "Best Effort" and "Persistent", with the first being the default. When a device is low on storage, the browser automatically tries to free up best effort storage. For instance, Firefox and Chromium-based browsers evict storage from the least recently used origins. This, however, poses a risk of losing critical data. To prevent eviction, developers can opt for persistent storage. In this case, the browser will not clear the storage, even when space is low. Users can still choose to clear up the storage manually. To opt for persistent storage, developers need to call the `navigator.storage.persist()` method. Depending on the browser and site engagement, a permission prompt may show, or the request will automatically be accepted or denied.

{{ figure_markup(
  image="storage_manager_api_persist.png",

>   alt="Percent of top pages using the persist method of the StorageManager API",
>   caption="Percent of top pages using the persist method of the StorageManager API.",

Reviewer: ```suggestion
  alt="Number of top pages using the persist method of the StorageManager API",
  caption="Number of top pages using the persist method of the StorageManager API.",
```

  description="Chart of the usage of StorageManager API's persist method, based on the number of pages monitored by HTTPArchive. It compares the usage on mobile and desktop devices. On desktop pages, the usage is almost steady, while there's more fluctuation on mobile devices. In October 2020, 25 desktop pages and 176 mobile pages make use of the API.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTxqot9ALgxcgOVJntkzIKnkpo3idIPy-tL0t_nzC5BwFuq0ThgK5OXOYVVOpama4vB2EyggX813d33/pubchart?oid=644836316&format=interactive",
  sheets_gid="1095648844",
  sql_file="durable_storage_persist_usage.sql"
  )
}}

> The `persist()` API is less often called than the `estimate()` method. Only 176 mobile pages make use of this API, compared to 25 desktop websites. While usage on the desktop seems to remain at this low level, there is no clear trend on mobile devices.

Author: Yes, it would be cool if you could assist me here.

* I've checked [the sample URLs here](https://chromestatus.com/metrics/feature/timeline/popularity/1369). The significantly higher number of mobile sites could be due to the classified ad provider Locanto, which uses the API on different subdomains, all of which count as separate sites.
* Apart from that, using the persist() API on mobile devices probably makes more sense, as they tend to have a smaller storage capacity compared to desktop devices.

{{ figure_markup(
  image="badging_api.png",
  alt="Percentage of page loads in Chrome using Badging API",
  caption='Percentage of page loads in Chrome using Badging API. (Source: <a href="https://chromestatus.com/metrics/feature/timeline/popularity/2726">Chrome Platform Status metrics for Badge Set</a>, <a href="https://chromestatus.com/metrics/feature/timeline/popularity/2727">Chrome Platform Status metrics for Badge Clear</a>)',
  description="Chart of Badging API usage, based on the percentage of page loads in Chrome using this feature. It compares the set and clear methods. The usage of both methods is growing over time, with the set method generally being called more often. In October 2020, there's a sudden growth for both methods, peaking at 0.025 percent of page loads for the set method and 0.016 percent for clear.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTxqot9ALgxcgOVJntkzIKnkpo3idIPy-tL0t_nzC5BwFuq0ThgK5OXOYVVOpama4vB2EyggX813d33/pubchart?oid=1145004925&format=interactive",
  sheets_gid="1154751352"
  )
}}

> In April 2020, Google Chrome 81 shipped the new Badging API, followed by Microsoft Edge 84 in July. After Chrome shipped the API, the usage numbers shot up. In October 2020, on 0.025 percent of all page loads in Google Chrome, the `setAppBadge()` method is called. The `clearAppBadge()` method is less often called, during around 0.016 percent of page loads.

Reviewer 2: Absolute page counts

Also, you mentioned that Twitter uses it. It could be insightful to query for all the URLs using it (or browse the sample URLs on the Chrome Status page itself) to find ones that are used on particularly large sites. Do you also see that this API has high usage in the % page views chart? That stat might be useful for comparison to make the point that few sites are using it but the ones that do are very large.

Author: Absolute page counts are not available (page loads), I'll have a look at the rest

Author: The stats use Chrome Platform Status data again (where we neither know the exact sites nor have absolute page counts available). HTTPArchive has only data available for the past three months. Both show a sudden growth in October. I don't see Twitter in the sample URLs though. The sample URLs seem to show various forums. It of course makes sense to present a badge if you decide to add the forum site to your homescreen. How should we proceed here?

### Notification Triggers API

The Push API requires the user to be online to receive a notification. Some applications, such as games, reminder or to-do apps, calendars, or alarm clocks, could also determine the target date for a notification locally and schedule it. To support this feature, the Chrome team experiments with a new API called [Notification Triggers](https://web.dev/notification-triggers/) ([Explainer](https://github.com/beverloo/notification-triggers/blob/master/README.md), not on a standards track yet). This API adds a new property called `showTrigger` to the `options` map that can be passed to the `showNotification()` method on the Service Worker's registration. The API is designed to allow for different kinds of triggers in the future, albeit for now, only time-based triggers are implemented. For scheduling a notification based on a certain date and time, developers can create a new instance of a `TimestampTrigger` and pass the target timestamp to it:

> ```js

Reviewer 2: (Side note: @barrypollard it'd be nice to support syntax highlighting)

Reviewer: We do support a limited form and use it in the methodology chapter but it basically needs to be lovely written by hand at the moment:

https://github.com/HTTPArchive/almanac.httparchive.org/blob/a5cbe77943ab7202c9ed55fdef71b20560a62033/src/templates/en/2020/methodology.html#L88-L104

That may be sufficient for now but a nicer code formatter would be nicer alright.

registration.showNotification('Title', {
  body: 'Message',
  showTrigger: new TimestampTrigger(timestamp),
});
```

{{ figure_markup(
  image="notification_triggers_api.png",
  alt="Percentage of page loads in Chrome using Notification Triggers API",
  caption='Percentage of page loads in Chrome using Notification Triggers API. (Source: <a href="https://chromestatus.com/metrics/feature/timeline/popularity/3017">Chrome Platform Status metrics</a>)',
  description="Chart of Notification Triggers API usage, based on the percentage of page loads in Chrome using this feature. It shows a peak in March 2020 with approximately 0.00003 percent of page loads, dropping to zero in October 2020.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTxqot9ALgxcgOVJntkzIKnkpo3idIPy-tL0t_nzC5BwFuq0ThgK5OXOYVVOpama4vB2EyggX813d33/pubchart?oid=1388597384&format=interactive",
  sheets_gid="1740370570"
  )
}}

> The Fugu team first experimented with Notification Triggers in an origin trial from Chrome 80 to 83, pausing development afterwards due to the lack of feedback by developers. Starting from Chrome 86 released in October 2020, the API has entered the origin trial phase again. This also explains the usage data of Notification Triggers API that peaked at being called on 0.000032 percent of page loads in Chrome during the first origin trial at around March 2020.

Reviewer 2: Absolute page counts

Author: Not available for Notification Triggers (page loads)

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTxqot9ALgxcgOVJntkzIKnkpo3idIPy-tL0t_nzC5BwFuq0ThgK5OXOYVVOpama4vB2EyggX813d33/pubchart?oid=1571330893&format=interactive",
  sheets_gid="708893754"
  )
}}

The use of the interface is currently still so low that it's hardly measurable. In October 2020, it has increased strongly and is now used during 0.00089 percent of page loads in Chrome.

## Conclusion

The state of web capabilities in 2020 is healthy, as new powerful APIs regularly ship with new releases of Chromium-based browsers. Some interfaces like the Content Indexing API or Idle Detection API help to add finishing touches to certain web applications. Other APIs, such as the File System Access and Async Clipboard API, allow a whole new application category, namely productivity apps, to finally fully make the shift to the web. Some APIs such as Async Clipboard and Web Share API have already made their way into other, non-Chromium browsers. Safari even was the first mobile browser to implement the Web Share API.

> Through its [rigorous process](https://developers.google.com/web/updates/capabilities#process), the Fugu team ensures that access to these features takes place in a secure and privacy-friendly manner. Additionally, the Fugu team actively solicits the feedback from other browser vendors and web developers. While the usage of most of these new APIs is comparatively low, some APIs presented in this chapter show an exponential or even hockey stick-like growth, such as the Badging or Content Indexing API. It is dependent on web developers what the state of web capabilities in 2021 will look like: The author encourages the community to build great web applications, make use of the powerful APIs in a backwards-compatible manner, and help make the web a more capable platform.

Reviewer 2: Could we include a link for the place where developers can give feedback on these APIs?

> # Capabilities

Reviewer 2: The title is auto-generated based on YAML metadata, which also includes info for authors, reviewers, analysts, etc.

Please see https://raw.githubusercontent.com/HTTPArchive/almanac.httparchive.org/main/src/content/en/2019/performance.md for an example of the metadata to include.

cc @bazzadp

Author: Almost done, I just need the ID of the new Discourse thread for this chapter.

Reviewer: Let’s use the same id as last year for now.

Author: Capabilities is a new chapter, so there's no thread from last year.

Reviewer: Ah silly me. Go with 1981 for now which is the 2020 announcement.

## Async Clipboard API

With the help of the `document.execCommand()` method, websites could already access the user's
clipboard. However, this approach is somewhat restricted, as the API is synchronous (making it
difficult to process clipboard items), and it can only interact with selected text in the DOM. This
is where the [Async Clipboard API](https://web.dev/async-clipboard/)
([W3C Working Draft](https://www.w3.org/TR/clipboard-apis/#async-clipboard-api)) comes in. This new
API is not only asynchronous, meaning it doesn't block the page for large chunks of data or waiting
for a permission to be granted, but also allows for images to be copied to or pasted from the
clipboard in supported browsers such as Chrome, Edge, and Safari.

### Read Access

The Async Clipboard API provides two methods for reading content from the clipboard: A shorthand
method for plain text, called `navigator.clipboard.readText()`, and a method for arbitrary data,
called `navigator.clipboard.read()`. Currently, most browsers only support HTML content and PNG
images as additional data formats. Due to privacy reasons, reading from the clipboard always
requires the user's consent.

> (IMAGE: Async Clipboard API)

Reviewer 2: For any visual or tabular figures to be embedded in the chapter, please refer to the [Figures Guide](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Figures-Guide) in the wiki for info on the expected markdown syntax to use and how to add metadata.

Author: Done, left the description at "TODO" until we decide which data to choose

---
part_number: II
chapter_number: 13
title: Capabilities
description: Capabilities chapter of the 2020 Web Almanac covering brand-new, powerful web platform APIs.
authors: [christianliebel]
reviewers: [logicalphase,tomayac]
analysts: [tomayac]
translators: []
discuss: 0
results: https://docs.google.com/spreadsheets/d/1N6j1qKv7vc51p1W9z_RrbJX9Se3Pmb-Uersfz4gWqqs/
queries: 13_Capabilities
published: 2020-11-07T00:00:00.000Z
last_updated: 2020-11-07T00:00:00.000Z
---

## Introduction

> The Capabilities Project, informally also known as Project Fugu, is a cross-company effort by
> Google, Microsoft, and Intel to close the gap between native applications and web apps. To do so,
> the Chromium contributors implement new APIs exposing capabilities of the operating system to the
> web, while maintaining user security, privacy, and trust. These capabilities include, but are not
> limited to:

Reviewer: Paragraphs should be written on one line as makes it easier to amend and also translate without having to reorder the lines completely and show more change than really is.

```suggestion
The Capabilities Project, informally also known as Project Fugu, is a cross-company effort by Google, Microsoft, and Intel to close the gap between native applications and web apps. To do so, the Chromium contributors implement new APIs exposing capabilities of the operating system to the web, while maintaining user security, privacy, and trust. These capabilities include, but are not limited to:
```

difficult to process clipboard items), and it can only interact with selected text in the DOM. This
is where the [Async Clipboard API](https://web.dev/async-clipboard/)
([W3C Working Draft](https://www.w3.org/TR/clipboard-apis/#async-clipboard-api)) comes in. This new
API is not only asynchronous, meaning it doesn't block the page for large chunks of data or waiting
for a permission to be granted, but also allows for images to be copied to or pasted from the
clipboard in supported browsers such as Chrome, Edge, and Safari.

### Read Access

The Async Clipboard API provides two methods for reading content from the clipboard: A shorthand
method for plain text, called `navigator.clipboard.readText()`, and a method for arbitrary data,
called `navigator.clipboard.read()`. Currently, most browsers only support HTML content and PNG
images as additional data formats. Due to privacy reasons, reading from the clipboard always
requires the user's consent.

(IMAGE: Async Clipboard API)

The Async Clipboard API is comparatively new, so its usage is currently rather low. In March 2020,
Safari added support for the Async Clipboard API in Safari 13.1. Over the course of 2020, the usage

> of the `read()` API was growing. In October 2020, the API was called during 0.0003 percent of all

Reviewer: In general we do encourage percentage stats rather than absolute stats, however I wonder should we make an exception here? 0.0003 percent of sites is a difficult to comprehend number (except that it's really small!) so should we quantify this with an actual number:

```suggestion
of the `read()` API was growing. In October 2020, the API was called during 0.0003 percent of all page loads in Google Chrome (or approximately 20 pages).
```

Or something like that. What do you think @rviscomi ?

Reviewer 2: If most of the stats are this small, I'd recommend a paragraph to set expectations with readers like:

> Many of the capabilities discussed in this chapter are still very new. So, unlike most chapters, stats will be presented as the absolute numbers of pages rather than relative percentages.

Author: @rviscomi Yes, most of the stats are this small. We need to check if we have absolute numbers for the Chrome Platform Status metrics available though.

Reviewer 2: Sorry it's taken me so long to understand that these numbers are from Chrome Status and not HTTP Archive. That being the case, only relative stats are available.

For the HTTP Archive-based stats like StorageManager API :: Estimate, we should switch over from percent of pages to number of pages for consistency with the other HA-based stats that have even smaller adoption and need to use absolute page counts. I think it's ok for HA and Chrome-based stats to be different in terms of % of **page views** and number of **pages**.
